Support fine grained activation sharding.

ptoulme-aws opened the pull request:
Hello, this is a draft PR to support fine grained activation sharding. All specs are defaulted to None, so this PR maintains all existing behavior of the Axlearn codebase.

I am requesting feedback and comments on the approach. 

With this PR the user can control more closely how GSPMD partitions activations allowing more control of the collectives generated by the SPMD partitioner. For example, GSPMD will generate all-to-alls around the gather operation of the vocab table, but with this PR user can annotate partition specs to generate a reduce-scatter instead.

Thank you!

ruomingp commented on axlearn/common/attention.py:
```diff
@@ -2771,11 +2788,14 @@ def _linear2(x):
         if cfg.structure == "prenorm":
+            inputs = maybe_shard(inputs, cfg.prenorm_partition_spec)
             x = self.norm(inputs)
+            x = maybe_shard(x, cfg.premlp_partition_spec)
```
Do we need this? I suppose `norm` usually does not change the partition spec?

ptoulme-aws replied:
We want to change the spec post norm to not propagate the spec from norm to MLP linear1. 
So for sequence parallel we would make the premlp_partition_spec = ((fsdp, data), None, None) to force an all-gather on the sequence dimension.

ruomingp replied:
Thanks for the explanation. Maybe we can add `output_partition_spec` to RMSNorm.Config to be consistent with Linear.Config. This is also more flexible as RMSNorm can be used in other places. WDYT?

ptoulme-aws replied:
Yes, I actually originally had this! I have revised to take this approach.

ruomingp commented on axlearn/common/attention.py:
```diff
@@ -2355,6 +2356,12 @@ class Config(BaseLayer.Config):
+        # If not None, how to partition pre attention activation values.
+        preattention_partition_spec: Optional[tuple[Optional[str]]] = None
+        # If not None, how to partition post attention activation values.
+        postattention_partition_spec: Optional[tuple[Optional[str]]] = None
```
How is this different from setting https://github.com/apple/axlearn/blob/2134a258a869f679e4dd2625565a2ca693126495/axlearn/common/attention.py#L3319?

kelvin-zou replied:
+1, I also have a high level question, should the sharding config auto-derived and optimized by XLA for most of the intermediary activations, based on weights and outputs etc? 
I'd like to see a concrete example to understand why explicit sharding is needed here.

ptoulme-aws replied:
OpenXLA ShardingPropagation often makes non optimal decisions in propagation. SPMD partitioner then partitions based on what HloSharding was propagated to each HloInstruction, so if your sharding propagation is not optimal then you will see collectives generated such as all-to-all to resolve the sharding conflicts.

An example is if you set the partition spec for premlp or preattention to be ((fsdp, data), model, None) to force the computation to be done in sequence parallel, then ShardingPropagation pass will propagate that HloSharding spec down to the matmul causing an all-to-all collective. 

This is not as optimal as an all-gather of the sequence dimension before the matmul.

ruomingp commented on axlearn/common/attention.py:
```diff
@@ -2355,6 +2356,12 @@ class Config(BaseLayer.Config):
         # Ref: https://github.com/google/praxis/blob/main/praxis/layers/transformers.py#L1129
         # TODO (bwzhang@) Adding a unittest for the hybridnorm.
         structure: str = "prenorm"
+        # If not None, how to partition pre norm activation values.
+        prenorm_partition_spec: Optional[tuple[Optional[str]]] = None
```
Please see my comment about `input_partition_spec` below.

ptoulme-aws replied:
Will revise to fix this.

ruomingp commented on axlearn/common/attention.py:
```diff
@@ -2355,6 +2356,12 @@ class Config(BaseLayer.Config):
+        # If not None, how to partition pre norm activation values.
+        prenorm_partition_spec: Optional[tuple[Optional[str]]] = None
+        # If not None, how to partition pre attention activation values.
+        preattention_partition_spec: Optional[tuple[Optional[str]]] = None
```
ditto.

ptoulme-aws replied:
Will revise to fix this.

ruomingp commented on axlearn/common/attention.py:
```diff
@@ -3315,6 +3335,21 @@ def set_ffn_partition_specs(ff_layer: TransformerFeedForwardLayer.Config):
+            preattention_partition_spec = (fsdp_axis_names, None, None)
+            postattention_partition_spec = (fsdp_axis_names, tp_axis_names, None)
+
+            layer_cfg.self_attention.set(
+                prenorm_partition_spec=prenorm_partition_spec, 
+                preattention_partition_spec=preattention_partition_spec, 
+                postattention_partition_spec=postattention_partition_spec)
+            layer_cfg.feed_forward.set(
+                prenorm_partition_spec=prenorm_partition_spec, 
+                premlp_partition_spec=preattention_partition_spec, 
+                postmlp_partition_spec=postattention_partition_spec)
```
Is `postattention_partition_spec` always the same as `prenorm_partition_spec` for `feed_forward` since the outputs of atten are used as inputs for ffn?

ptoulme-aws replied:
It is; which one would you prefer to keep or rename both to input_partition_spec as you suggested?

ruomingp reviewed (changes requested):
Thanks! Some questions about the API...

kelvin-zou commented on axlearn/common/layers.py:
```diff
@@ -814,8 +821,14 @@ def _create_layer_parameter_specs(self) -> dict[str, ParameterSpec]:
         )
 
     def forward(self, x: Tensor) -> Tensor:
+        cfg = self.config
```
Same question here, can we handle this inside XLA compiler directly instead of making it explicit here? Adding this kind of fine-level sharding spec is a divergence from GSPMD programming paradigm imho.

ptoulme-aws replied:
Responded above with a concrete example. The main issues we are trying to solve here are 

1. We want Norms/Dropouts/Residuals to be computed in sequence parallel so partitioned along sequence dimension. This is to avoid redundant compute across tensor parallel workers.
2.. OpenXLA ShardingPropagation often makes non optimal decisions - [https://github.com/openxla/xla/blob/main/xla/service/sharding_propagation.h]([https://github.com/openxla/xla/blob/main/xla/service/sharding_propagation.h])
3. Those non optimal decisions such as say propagating sequence level sharding to the left hand side of the attention QKV proj or MLP up proj will cause the SPMD partitioner to insert all-to-alls or collective-permutes or involuntary full rematerialization.


I am curious, why Axlearn has an activation level sharding spec in use today for `linear2.output_partition_spec`
Was this added to also address non optimal SPMD partitioning or propagation?

kelvin-zou commented on axlearn/common/attention_test.py:
```diff
@@ -5091,6 +5091,70 @@ def test_set_double_shard_weights_config_for_list_of_configs(
+        cfg_layer.self_attention.attention.input_linear = self_attention_input_linear_cfg
+        cfg_layers = [cfg_layer, cfg_layer]
+
+        cfg_layer.self_attention.prenorm_partition_spec = (fsdp_axis_names, tp_axis_names, None,)
```
interesting, is it the real use case you are testing nowadays?

ptoulme-aws replied:
Yes, this is the real config for sequence parallel. I have confirmed with these specs and internal Neuron changes we can generate partitioned HLOs that perform Norms/Residuals/Dropouts partitioned along the sequence and matmuls partitioned along the model dimension without any unnecessary collectives, such as all-to-alls or collective-permutes or involuntary full rematerialization.

ptoulme-aws commented on axlearn/common/attention.py:
```diff
@@ -3315,6 +3335,21 @@ def set_ffn_partition_specs(ff_layer: TransformerFeedForwardLayer.Config):
             set_ffn_partition_specs(layer_cfg.feed_forward)
+            
+        # Neuron backend needs fine grained activation sharding.
+        if jax.default_backend() == 'neuron':
```
Got it. I will revise to make this a separate function instead.

ptoulme-aws commented on axlearn/common/attention.py:
```diff
@@ -2708,6 +2718,13 @@ class Config(BaseLayer.Config):
+        # If not None, how to partition pre MLP activation values.
+        premlp_partition_spec: Optional[tuple[Optional[str]]] = None
+        # If not None, how to partition post MLP activation values.
+        postmlp_partition_spec: Optional[tuple[Optional[str]]] = None
```
Good catch! This one is not needed.

ptoulme-aws commented on axlearn/common/attention.py:
```diff
@@ -2771,11 +2788,14 @@ def _linear2(x):
 
         remat_pt2 = "linear2"
         if cfg.structure == "prenorm":
+            inputs = maybe_shard(inputs, cfg.prenorm_partition_spec)
```
Will revise to take this approach.

ruomingp commented:
Thanks. Please re-request review when it's ready.

ptoulme-aws commented on axlearn/common/utils.py:
```diff
@@ -444,6 +444,12 @@ def with_sharding_constraint(x, shardings):
+        return x
+    return with_sharding_constraint(x, PartitionSpec(*partition_spec))
+
+
```
pre-commit wanted these blank lines. I ran 

`pre-commit run -a`

ptoulme-aws commented:
@kelvin-zou  @ruomingp 

I have revised the PR based on your comments. I removed "WIP," as the PR is now production ready. I removed Neuron specific logic; I think that should be in a separate PR, and this PR should be enabling hardware agnostic activation sharding. 

I ran precommit and pytest and both are passing. Please take a look! Thank you.

ruomingp reviewed (commented):
Thanks! Are there tests that can be added for the new logic?

ptoulme-aws commented:
@ruomingp What test do you envision? What should I assert? This is effectively just putting 

with_sharding_constraint(tensor, PartitionSpec('fdsp', 'model')) 

behind a config.

ptoulme-aws commented:
@ruomingp  I could add a test where I make an RMSNorm and set the config for partition specs and then assert they are the same specs, but I am not sure what benefit that adds. I do not see tests like that for other configs, as that is like a setter/getter test.

ptoulme-aws commented on axlearn/common/utils.py:
```diff
@@ -444,6 +444,12 @@ def with_sharding_constraint(x, shardings):
     return jax.lax.with_sharding_constraint(x, shardings)
 
 
+def maybe_shard(x, partition_spec) -> Tensor:
```
x is not always a tensor. For example, with `aten_output `from attention it is not a tensor. I can add the type for partition_spec

ruomingp commented:
> @ruomingp I could add a test where I make an RMSNorm and set the config for partition specs and then assert they are the same specs, but I am not sure what benefit that adds. I do not see tests like that for other configs, as that is like a setter/getter test.

How about patching `with_sharding_constraint` to make sure that it's called with the right tensors in forward() when we set partition specs?

ruomingp commented on axlearn/common/utils.py:
```diff
@@ -443,6 +443,11 @@ def with_sharding_constraint(x, shardings):
         return x
     return jax.lax.with_sharding_constraint(x, shardings)
 
+def maybe_shard(x: Tensor, partition_spec: Optional[PartitionSpec]) -> Tensor:
```
Is `x` a `Nested[Tensor]`?

ptoulme-aws replied:
I think sometimes it is. That is why I did not have a type. I know for `aten_output `it is not a tensor. 

What type do you recommend?

ptoulme-aws commented:
> > @ruomingp I could add a test where I make an RMSNorm and set the config for partition specs and then assert they are the same specs, but I am not sure what benefit that adds. I do not see tests like that for other configs, as that is like a setter/getter test.
> 
> How about patching `with_sharding_constraint` to make sure that it's called with the right tensors in forward() when we set partition specs?

Do you have an example of this?

ptoulme-aws commented:
@ruomingp @kelvin-zou 
I have added tests per @ruomingp specifications. All tests pass and pre-commit passes also. 

Thank you!
